solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+            revert MalformedCurrentValidatorSet();
+        }
+
+        // Check that the supplied current validator set matches the saved checkpoint
+        bytes32 currentValsetHash = keccak256(abi.encode(_currentValidatorSet));
+        if (
+            domainSeparateValidatorSetHash(BRIDGE_ID, currentNonce, currentPowerThreshold, currentValsetHash) !=
+            s_lastValidatorSetCheckpoint
+        ) {
+            revert SuppliedValidatorSetInvalid();
+        }
+
+        // Check that enough current validators have signed off on the new validator set
+        bytes32 newCheckpoint = domainSeparateValidatorSetHash(
+            BRIDGE_ID,
+            currentNonce,
+            _newPowerThreshold,
+            _newValidatorSetHash
+        );
+        checkValidatorSignatures(_currentValidatorSet, _sigs, newCheckpoint, currentPowerThreshold);
+
+        // EFFECTS
+
+        s_lastValidatorSetCheckpoint = newCheckpoint;
+        s_powerThreshold = _newPowerThreshold;
+        s_lastValidatorSetNonce = _newNonce;
+
+        // LOGS
+
+        emit ValidatorSetUpdatedEvent(_newNonce, _newPowerThreshold, _newValidatorSetHash);
+    }
+
+    /// @notice Relays a batch of Celestia -> Ethereum messages. Anyone can
+    // call this function, but they must supply valid signatures of the current
+    // validator set over the batch.
+    /// @param _nonce The message root nonce.
+    /// @param _messageRoot The Merkle root of messages.
+    /// @param _currentValidatorSet The current validator set.
+    /// @param _sigs Signatures.
+    function submitMessageRoot(
+        uint256 _nonce,
+        bytes32 _messageRoot,
+        Validator[] calldata _currentValidatorSet,
+        Signature[] calldata _sigs
+    ) external {
+        // CHECKS
+
+        uint256 currentPowerThreshold = s_powerThreshold;
+
+        // Check that the batch nonce is higher than the last nonce for this token
+        if (_nonce <= s_lastMessageRootNonce) {
+            revert InvalidMessageRootNonce();
+        }
+
+        // Check that current validators and signatures are well-formed.
+        if (_currentValidatorSet.length != _sigs.length) {
+            revert MalformedCurrentValidatorSet();
+        }
+
+        // Check that the supplied current validator set matches the saved checkpoint
+        bytes32 currentValsetHash = keccak256(abi.encode(_currentValidatorSet));
+        if (
+            domainSeparateValidatorSetHash(
+                BRIDGE_ID,
+                s_lastValidatorSetNonce,
+                currentPowerThreshold,
+                currentValsetHash
+            ) != s_lastValidatorSetCheckpoint
+        ) {
+            revert SuppliedValidatorSetInvalid();
+        }
+
+        // Check that enough current validators have signed off on the message root and nonce
+        bytes32 c = domainSeparateMessageRoot(BRIDGE_ID, _nonce, _messageRoot);
+        checkValidatorSignatures(_currentValidatorSet, _sigs, c, currentPowerThreshold);
```

Comment:
The contract does not impose any expectations about the message root. It can literally be any 32 bytes string. Shouldn't there be a verification logic that these 32 bytes represent the root of a Merkle tree of certain leaves?

Reply:
Not here. Under an honest majority assumption, the root will be correctly constructed.

Reply:
> There separately needs to be Merkle proof verification logic, but I'll save that for a future PR.

Can you please open an issue and describe the proof verification logic + required changes there, please?

Reply:
#6